ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -0,0 +1,192 @@
+                     tabulate, (===))
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+
+import           Control.Monad.Class.MonadAsync (MonadAsync)
+import           Control.Monad.Class.MonadFork (MonadFork)
+import           Control.Monad.Class.MonadSTM (MonadSTM, STM, atomically)
+import           Control.Monad.Class.MonadThrow (MonadMask, MonadThrow)
+import           Control.Monad.IOSim (runSimOrThrow)
+
+import           Ouroboros.Consensus.Ledger.Abstract (ledgerConfigView)
+import           Ouroboros.Consensus.Mempool (Mempool (..),
+                     MempoolSnapshot (..), openMempool)
+import           Ouroboros.Consensus.Mempool.TxSeq (TicketNo)
+import           Ouroboros.Consensus.Util.ThreadRegistry (withThreadRegistry)
+
+import           Ouroboros.Network.Chain (Chain (..))
+
+import           Ouroboros.Storage.ChainDB.API (ChainDB)
+import qualified Ouroboros.Storage.ChainDB.API as ChainDB
+import qualified Ouroboros.Storage.ChainDB.Mock as Mock
+
+import           Test.Util.TestBlock (GenTx (..), GenTxId (..), TestBlock,
+                     computeGenTxId, singleNodeTestConfig, testInitExtLedger)
+import           Test.Util.TestTx (TestTx (..))
+
+tests :: TestTree
+tests = testGroup "Mempool"
+  [ testProperty "getTxs == getTxsAfter zeroIdx"        prop_Mempool_getTxs_getTxsAfter
+  , testProperty "addedTxs == getTxs"                   prop_Mempool_addTxs_getTxs
+  , testProperty "Invalid transactions are never added" prop_Mempool_InvalidTxsNeverAdded
+  ]
+
+{-------------------------------------------------------------------------------
+  Mempool Implementation Properties
+-------------------------------------------------------------------------------}
+
+-- | Test that @getTxs == getTxsAfter zeroIdx@.
+prop_Mempool_getTxs_getTxsAfter :: Chain TestBlock
+                                -> [TestTx]
+                                -> Property
+prop_Mempool_getTxs_getTxsAfter bc txs =
+  testAddTxsWithMempoolAndSnapshot
+    bc
+    txs
+    (\Mempool{zeroIdx} MempoolSnapshot{getTxs, getTxsAfter} ->
+      getTxs === getTxsAfter zeroIdx)
+
+-- | Test that all valid transactions added to a 'Mempool' can be retrieved
+-- afterward.
+--
+-- n.b. In this test, we add a list of arbitrary transactions to an initially
+-- empty 'Mempool', retrieve all of the transactions from the 'Mempool', and
+-- assert that those transactions are /specifically/ equal to the /valid/
+-- transactions that we attempted to add (the invalid ones should have been
+-- dropped).
+prop_Mempool_addTxs_getTxs :: Chain TestBlock
+                           -> [TestTx]
+                           -> Property
+prop_Mempool_addTxs_getTxs bc txs =
+    testAddTxsWithMempoolAndSnapshot
+        bc
+        txs
+        (\_ MempoolSnapshot{getTxs} ->
+          filter (genTxIsValid . snd) (testTxsToGenTxPairs txs)
+              === dropThd getTxs)
+  where
+    dropThd :: [(a, b, c)] -> [(a, b)]
```

Comment:
Tuples with a size > 2 always make me think: shouldn't we turn it into a record so that we can use an accessor function instead of `dropThd`?

The reason for these 3-tuples is:
```haskell
data MempoolSnapshot txid tx idx = MempoolSnapshot {
    getTxs :: [(txid, tx, idx)]
  , ..
}
```

What if we define it as:
```haskell
data MempoolSnapshot blk idx = MempoolSnapshot {
    getTxs :: [(GenTx blk, idx)]
  , ..
}
```
Then we have `computeGenTxId` to get the `txid` of `tx` and are back to a 2-tuple. In general, I like to be as abstract as possible, but sometimes being more specific allows us to be cleverer and less redundant, like in this case.

What do you think about this change? If you agree, this is of course something for a separate PR.

Reply:
I think this is a good idea provided that I make the `computeGenTxId` function a bit cheaper than it currently is (as it stands, this function would compute the transaction hash every time it's called).

Duncan suggested that I employ some kind of memoization or caching functionality and rename it to `getGenTxId` (it was called `computeGenTxId` that it's not a necessarily cheap operation).

Reply:
> I think this is a good idea provided that I make the `computeGenTxId` function a bit cheaper than it currently is (as it stands, this function would compute the transaction hash every time it's called).

Indeed, if it needs to compute the hash every time it's called, it's not cheap enough.

> Duncan suggested that I employ some kind of memoization or caching functionality and rename it to `getGenTxId` (it was named `computeGenTxId` to imply that it's not a necessarily cheap operation).

That sounds good.